extend.sh

```diff
@@ -0,0 +1,466 @@
+        exit 1
+    fi
+}
+
+function get_device_type(){
+    val=$( /usr/sbin/pvs --noheadings -o NAME,fmt | /usr/bin/sed -n 's#'"$DEVICE_NAME""$ADJACENT_PARTITION_NUMBER"' ##p' | /usr/bin/awk 'NF { $1=$1; print }' 2>&1)
+    local status=$?
+    if [[ status -ne 0 ]]; then
+        echo "Failed to retrieve device type: $val"
+        exit 1
+    fi
+    echo "$val"
+}
+
+function ensure_fs_is_ext4(){
+    local device=$1
+    fstype=$(/usr/bin/lsblk -fs "$device" --noheadings -o FSTYPE -d| /usr/bin/awk 'NF { $1=$1; print }')
+    local status=$?
+    if [[ $status -ne 0 ]]; then
+        exit $status
+    fi
+    if [[ "$fstype" != "ext4" ]]; then
+        echo "Device $device does not contain an ext4 file system: $fstype"
+        exit 1
+    fi
+}
+
+function resolve_device_name(){
+    local device=$DEVICE_NAME$ADJACENT_PARTITION_NUMBER
+    device_type=$(get_device_type "$device")
+    if [[ $device_type == "lvm2" ]]; then
+        # It's an LVM block device 
+        # Determine which is the last LV in the PV
+        # shellcheck disable=SC2016
+        device=$(/usr/sbin/pvdisplay "$device" -m | /usr/bin/sed  -n '/Logical volume/h; ${x;p;}' | /usr/bin/awk  '{print $3}')
+        status=$?
+        if [[ status -ne 0 ]]; then
+            echo "Failed to identify the last LV in $device"
+            exit $status
+        fi
+        # Capture the LV device name
+        LOGICAL_VOLUME_DEVICE_NAME=$device
+    fi
+}
+
+
+function deactivate_volume_group(){
+    local volume_group_name
+    volume_group_name=$(/usr/sbin/pvs "$DEVICE_NAME""$ADJACENT_PARTITION_NUMBER" -o vg_name --noheadings|/usr/bin/sed 's/^[[:space:]]*//g')
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        echo "Failed to retrieve volume group name for logical volume $LOGICAL_VOLUME_DEVICE_NAME: $err"
+        exit $status
+    fi
+    ret=$(/usr/sbin/vgchange -an "$volume_group_name" 2>&1)
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        echo "Failed to deactivate volume group $volume_group_name: $err"
+        exit $status
+    fi
+    # avoid potential deadlocks with udev rules before continuing
+    sleep 1
+}
+
+function check_device(){
+    local device=$DEVICE_NAME$ADJACENT_PARTITION_NUMBER
+    resolve_device_name
+    if [[ -n $LOGICAL_VOLUME_DEVICE_NAME ]]; then
+        #calculate_required_lv_shrinking_size
+        # if it's a logical volume then use the mapped device name
+        device=$LOGICAL_VOLUME_DEVICE_NAME
+    fi
+    ensure_device_not_mounted "$device"
+    ensure_fs_is_ext4 "$device"
+    check_filesystem "$device"
+    check_filesystem_size "$device"
+}
+
+function shrink_logical_volume() {
+    ret=$(/usr/sbin/lvreduce --resizefs -L -"$INCREMENT_BOOT_PARTITION_SIZE" "$LOGICAL_VOLUME_DEVICE_NAME" 2>&1)
+    local status=$?
+    if [[ $status -ne 0 ]]; then
+        echo "Failed to shrink logical volume $LOGICAL_VOLUME_DEVICE_NAME: $err"
```

Comment:
```suggestion
        echo "Failed to shrink logical volume $LOGICAL_VOLUME_DEVICE_NAME: $ret"
```
Same thing in a few other places in the script.